folly/external/aor/memchr-advsimd.S

```diff
@@ -0,0 +1,239 @@
+/*
+ * memchr - find a character in a memory zone
+ *
+ * Copyright (c) 2014-2022, Arm Limited.
+ * Copyright (c) 2025, Nvidia
+ * SPDX-License-Identifier: MIT OR Apache-2.0 WITH LLVM-exception
+ */
+
+/* Assumptions:
+ *
+ * ARMv8-a, AArch64
+ * Neon Available, sha-512 Available
+ */
+
+#include "asmdefs.h"
+
+.arch armv8-a+sha3
+
+/* Arguments and results.  */
+#define srcin           x0
+#define chrin           w1
+#define cntin           x2
+
+#define result          x0
+
+#define src             x3
+#define tmp             x4
+#define wtmp2           w5
+#define synd            x6
+#define soff            x7
+#define cntrem          x8
+#define end_addr        x9
+
+#define vrepchr         v0
+#define vdata1          v1
+#define vdata2          v2
+#define vdata3          v3
+#define vdata4          v4
+#define vdata5          v5
+#define vdata6          v6
+#define vdata7          v7
+#define vhas_chr1       v8
+#define vhas_chr1q      q8
+#define vhas_chr2       v9
+#define vhas_chr2q      q9
+#define vhas_chr3       v10
+#define vhas_chr4       v11
+#define vhas_chr4q      q11
+#define vhas_chr5       v12
+#define vhas_chr5q      q12
+#define vhas_chr6       v13
+#define vhas_chr7       v14
+#define vrepmask        v15
+#define vend            v16
+
+.macro  CHECK_PAIR regA, regB
+    cmeq    vhas_chr1.16b, \regA\().16b, vrepchr.16b
+    cmeq    vhas_chr2.16b, \regB\().16b, vrepchr.16b
+    and     vhas_chr1.16b, vhas_chr1.16b, vrepmask.16b
+    and     vhas_chr2.16b, vhas_chr2.16b, vrepmask.16b
+    addp    vend.16b, vhas_chr1.16b, vhas_chr2.16b
+    addp    vend.16b, vend.16b, vend.16b  // reduces 128 -> 64 bits
+    mov     synd, vend.d[0]
+    cbnz    synd, L(tail)
+.endm
+
+.macro  CHECK_SINGLE reg
+    cmeq    vhas_chr2.16b, \reg\().16b, vrepchr.16b
+    and     vhas_chr2.16b, vhas_chr2.16b, vrepmask.16b
+    // Here we addp the same register to itself, effectively computing OR 
+    // across lanes again, then reduce to 64 bits:
+    addp    vend.16b, vhas_chr2.16b, vhas_chr2.16b 
+    addp    vend.16b, vend.16b, vend.16b
+    mov     synd, vend.d[0]
+    cbnz    synd, L(tail)
+.endm
+
+/*
+ * Core algorithm:
+ *
+ * For each 32-byte chunk we calculate a 64-bit syndrome value, with two bits
+ * per byte. For each tuple, bit 0 is set if the relevant byte matched the
+ * requested character and bit 1 is not used (faster than using a 32bit
+ * syndrome). Since the bits in the syndrome reflect exactly the order in which
+ * things occur in the original string, counting trailing zeros allows to
+ * identify exactly which byte has matched. Unrolled to 112 bytes
+ * and using SHA512 instruction for main loop reduction - should only compile for
+ * devices with h/w SHA512 instruction support. MTE compatibility TBD
+ */
+
+ENTRY (__folly_memchr_aarch64_simd)
```

Comment:
Let's include a description of the variant in the name - this is a long-string variant - `__folly_memchr_long_aarch64_simd`.